dask/array/core.py

```diff
@@ -2981,12 +2981,28 @@ def concatenate(seq, axis=0, allow_unknown_chunksizes=False):
     """
     seq = [asarray(a) for a in seq]
 
-    n = len(seq)
-
-    if n == 0:
+    if not seq:
         raise ValueError("Need array(s) to concatenate")
 
+    # Coerce all arrays to the same type
+    # Empty arrays can impact the output dtype
+    from .utils import meta_from_array
+    metas = [getattr(s, '_meta', s) for s in seq]
+    metas = [meta_from_array(m, getattr(m, 'ndim', 1)) for m in metas]
```

Comment:
Is it safe to assume here that each `_meta` attribute has an `ndim` attribute? Why is the default `ndim` 1?

Reply:
This is just @pentschev's code moved up a few lines. Maybe he can answer your question?

That said, I don't really plan to change this in this PR as we are not introducing it; only moving it.

Reply:
@jrbourbeau here `seq` is a list of `dask.array`s, all `dask.array`s necessarily have the `_meta` attribute. `seq` can also contain `list`s of `dask.array`s, which will be treated recursively in `meta_from_array`. The `ndim` argument in `meta_from_array` (which isn't perfect yet) is currently mandatory, so that's why I'm defaulting it to `1`, but as of now, it would work just fine for any other values as well.

In the near future I intend to find a more stable implementation of `meta_from_array`, but it's not yet fully clear what are all the conditions we need to consider, so that's why I didn't give a lot of thought on a cleaner version. My main thought process for now is: make sure it's required to prevent accidental omission, which may be difficult to debug.

Reply:
I've gone ahead and pushed some simplifications to the `_meta` handling code here. Hopefully that addresses this concern. Let's follow-up on the new diff if not.

Reply:
Mostly I wanted to prevent you from drifting too much from your original work, but I don't mind if you take care of these things and reduce my workload. :D 

Thanks for doing that @jakirkham!